kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncThreadPool.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.api.async.internals;
```

Comment:
The thread pooling structure seems a little overly complex to me. Here's some problems I see with how this is structured in the pr:
1. we have to reimplement thread pools, which is a somewhat complex concurrent data structure (there's some bug-prone synchronization code, thread management, queueing)
2. we have this concept of tasks/partitions baked into the thread pool, which adds to the complexity
3. the context routing code is also quite complex and somewhat difficult to follow.

I think we can simplify that layer by doing the following:
1. build on top of the executor library in java. I think it does everything you need already outside of mass-canceling events associated with a closed task. I'll get to how to do that in a bit.
2. bake all the context loading into the runnable that gets sent to the executors, and load it using thread-locals instead of your own router

For cancelling events associated with a task - the executors already let you do this. You just need to hang on to the events so you can cancel them. Its more expensive, but we cant have that many of these outstanding at once anyway, and closes are pretty uncommon.

This might be hard to envision from a comment, so I put up an implementation of this approach over on this branch: https://github.com/responsivedev/responsive-pub/commit/a54e2f789e90312a070d4bf56554aa5727d68829.If you agree its simpler, then obviously just use it.